src/typescript-worker/util.ts

```diff
@@ -27,11 +32,23 @@ export class MergedAsyncIterables<T> {
       next,
     });
     this._iterators.set(next, next());
+    this._newIterator.resolve(NEW_ITERATOR);
   }
 
   async *[Symbol.asyncIterator]() {
     while (this._iterators.size > 0) {
-      const {value, next} = await Promise.race(this._iterators.values());
+      // Emit any available value from our source iterators, or notice there is
+      // a new iterator. We should immediately add new iterators to this race,
+      // because a new one might have a value before the older ones.
+      const result = await Promise.race([
+        ...this._iterators.values(),
```

Comment:
This definitely works and should go in, but I feel like it'd be a bit more straight-forward with an approach that iterates on each iterable directly. Essentially combining an async iteratable that let's you push new individual values in (like TestAsyncIterable in lit-html tests) with this...

```ts
add(iterable: AsyncIterable<T>) {
  for await (const i of iterable) {
    await this.push(i);
  }
}
```
...to push in new values. I haven't tested, but here's the idea:

```ts
export class MergedAsyncIterables<T> {
  private _buffer: Array<{value: T, emitted: () => void}> = [];  
  private _notifyNewValue?: () => void;
    
  async *[Symbol.asyncIterator]() {
    while (true) {
      if (this._buffer.length > 0) {
        const {value, emitted} = this._buffer.shift();
        yield value;
        // Let the loop in add() continue
        emitted();
      } else {
        // Wait for a new value
        await new Promise<void>((resolve) => (this._notifyNewValue = resolve));
        this._notifyNewValue = undefined;
      }
    }
  }
  
  async add(iterable: AsyncIterable<T>) {
    for await (const i of iterable) {
      // Wait for this value to be emitted before continuing
      await this._push(i);
    }
  }
  
  private async _push(value: T): Promise<void> {
    let emitted: () => void;
    const emittedPromise = new Promise<void>((res) => emitted = res);
    this._buffer.push({value, emitted});
    this._notifyNewValue?.();
    // Let the loop in add() wait for this value to be emitted
    return emittedPromise;
  }
}
```

Reply:
Nice! That is simpler.

The only change I had to make was keeping track of how many active source iterators there were, and also notifying when one is exhausted, otherwise we never know when we are done. Slightly modified some other code too, but only the one change to the logic.

PTAL